src/dmd/json.d

```diff
@@ -857,17 +871,178 @@ public:
 extern (C++) void json_generate(OutBuffer* buf, Modules* modules)
 {
     scope ToJsonVisitor json = new ToJsonVisitor(buf);
-    json.arrayStart();
-    json.generateCompilerInfo();
-    json.generateBuildInfo();
-    for (size_t i = 0; i < modules.dim; i++)
-    {
-        Module m = (*modules)[i];
-        if (global.params.verbose)
-            message("json gen %s", m.toChars());
-        m.accept(json);
-    }
-    json.generateSemantics();
-    json.arrayEnd();
-    json.removeComma();
+
+    if (global.params.jsonQueryFlags == 0)
+    {
+        // Generate the original "non-query format, which is just an array
+        // of modules representing their syntax.
+        json.generateModules(modules);
+        json.removeComma();
+    }
+    else
+    {
+        // Generate the "query format" which is an object where each
+        // output option is its own field.
+
+        json.objectStart();
+        if (global.params.jsonQueryFlags & JsonQueryFlags.compilerInfo)
+        {
+            json.propertyStart("compilerInfo");
+            json.generateCompilerInfo();
+        }
+        if (global.params.jsonQueryFlags & JsonQueryFlags.buildInfo)
+        {
+            json.propertyStart("buildInfo");
+            json.generateBuildInfo();
+        }
+        if (global.params.jsonQueryFlags & JsonQueryFlags.modules)
+        {
+            json.propertyStart("modules");
+            json.generateModules(modules);
+        }
+        if (global.params.jsonQueryFlags & JsonQueryFlags.semantics)
+        {
+            json.propertyStart("semantics");
+            json.generateSemantics();
+        }
+        json.objectEnd();
+    }
+}
+
+/**
+Each flag represents a field that can be included in the JSON output.
+*/
+private enum JsonQueryFlags
+{
+    compilerInfo = 0x01,
+    buildInfo    = 0x02,
+    modules      = 0x04,
+    semantics    = 0x08,
+}
+private enum jsonFieldNames = () {
+    string s;
+    string prefix = "";
+    foreach (enumName; __traits(allMembers, JsonQueryFlags))
+    {
+        s ~= prefix ~ enumName;
+        prefix = ", ";
+    }
+    s ~= "\0"; // make sure it is null-terminated
+    return s;
+}();
+
+/**
+Parses a JSON query into a set of flags.
+
+Returns: false on error, prints its own error messages.
+*/
+bool parseJsonQuery(const(char)* query, uint* outQueryFlags)
```

Comment:
`out uint outQueryFlags`

Reply:
or `ref` as it is cumulative.

Reply:
Ugh, I hate that now the caller won't know their "value" argument could be modified, i.e.
```D
ubyte myFlags = 0;
parseJsonQuery(myQuery, myFlags);
```
Now `myFlags` was modified, but it's not clear from the call that this happened.  I would prefer to see:
```D
ubyte myFlags = 0;
parseJsonQuery(myQuery, &myFlags);
```
Now you definitely know that `myFlags` could have been modified.  That's just my preference, but this is not my code so I'll do whatever the core team prefers for `dmd`.